mono/mini/mini-runtime.c

```diff
@@ -1905,6 +1905,157 @@ mono_jit_map_is_enabled (void)
 
 #endif
 
+#if defined(ENABLE_JIT_DUMP)
+static FILE* perf_dump_file;
+static mono_mutex_t perf_dump_mutex;
+static guint32 perf_dump_pid;
+
+enum {
+	JIT_DUMP_MAGIC = 0x4A695444,
+	JIT_DUMP_VERSION = 2,
+#if HOST_X86
+	ELF_MACHINE = EM_386,
+#elif HOST_AMD64
+	ELF_MACHINE = EM_X86_64,
+#elif HOST_ARM
+	ELF_MACHINE = EM_ARM,
+#elif HOST_ARM64
+	ELF_MACHINE = EM_AARCH64,
+#endif
+	JIT_CODE_LOAD = 0
+};
+typedef struct
+{
+	guint32 magic;
+	guint32 version;
+	guint32 total_size;
+	guint32 elf_mach;
+	guint32 pad1;
+	guint32 pid;
+	guint64 timestamp;
+	guint64 flags;
+}FileHeader;
+typedef struct
+{
+	guint32 id;
+	guint32 total_size;
+	guint64 timestamp;
+}RecordHeader;
+typedef struct
+{
+	RecordHeader header;
+	guint32 pid;
+	guint32 tid;
+	guint64 vma;
+	guint64 code_addr;
+	guint64 code_size;
+	guint64 code_index;
+	// Null terminated function name
+	// Native code
+}JitCodeLoadRecord;
+
+static void add_file_header_info (FileHeader *header);
+static guint64 get_time_stamp_ns (void);
+static void add_basic_JitCodeLoadRecord_info (JitCodeLoadRecord *record);
+
+void
+mono_enable_jit_dump (void)
+{
+	if (perf_dump_pid == 0)
+		perf_dump_pid = getpid();
+	
+	if (!perf_dump_file) {
+		char name [64];
+		FileHeader header;
+		memset (&header, 0, sizeof (FileHeader));
+
+		mono_os_mutex_init (&perf_dump_mutex);
+		mono_os_mutex_lock (&perf_dump_mutex);
+		
+		g_snprintf (name, sizeof (name), "/tmp/jit-%d.dump", perf_dump_pid);
+		unlink (name);
+		perf_dump_file = fopen (name, "w");
+		
+		add_file_header_info (&header);
+		fwrite (&header, sizeof (FileHeader), 1, perf_dump_file);
+		
+		mmap (NULL, sizeof (FileHeader), PROT_READ | PROT_EXEC, MAP_PRIVATE, 1, 0);
```

Comment:
`sizeof (header)`.
Check return value of `mmap`.

Reply:
Change 1 to `STDOUT_FILENO`?
This is weird though.

Reply:
@fanyang-mono Use `sizeof(header)` and `STDOUT_FILENO` as Jay suggested.  Also this line could really really use a comment (something like what you explained to @lpereira previously) that this is used to notify `perf` about the `jitdump`.  It's really non-obvious when casually reading the code.

Reply:
Hm. It looks like it shoudln't be `1` (stdout), it should be the `fileno (perf_dump_file)`

Compare the jitdump code in CoreCLR here: https://github.com/dotnet/coreclr/pull/26897/files#diff-05a5e90ba742e6131eea2114bbb4fadbR190-R211


Also @fanyang-mono you should save the address that you get back from `mmap` to a static global variable, and then `munmap` it in `mono_jit_dump_cleanup`

Reply:
Ah, being `fileno(perf_dump_file)` makes a whole lot more sense!

Reply:
I agree.